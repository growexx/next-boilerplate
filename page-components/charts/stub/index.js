export const getAPIMock = () => {
  const data = []
  for (let i = 0; i < 10; i++) {
    data.push({
      name: `${i} - Label`,
      count: Math.floor(Math.random() * 100)
    })
  }

  return Promise.resolve({
    data,
    status: 1
  })
}
