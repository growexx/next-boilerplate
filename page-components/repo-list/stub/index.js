export const TEST_IDS = {
  REPO_FORM: 'REPO_FORM',
  REPO_LOADER: 'REPO_LOADER'
}

export const mockRepoData = [
  {
    id: 2618963,
    node_id: 'MDEwOlJlcG9zaXRvcnkyNjE4OTYz',
    name: 'AFNetworking',
    full_name: 'js/AFNetworking',
    private: false,
    owner: {
      login: 'js',
      id: 23626
    }
  },
  {
    id: 459905452,
    node_id: 'R_kgDOG2mZrA',
    name: 'analytics-ios-integration-mixpanel',
    full_name: 'js/analytics-ios-integration-mixpanel',
    private: false,
    owner: {
      login: 'js',
      id: 23626
    }
  }
]
