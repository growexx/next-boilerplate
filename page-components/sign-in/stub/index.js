export const TEST_IDS = {
  GITHUB_SIGN_IN_BUTTON: 'GITHUB_SIGN_IN_BUTTON',
  GOOGLE_SIGN_IN_BUTTON: 'GOOGLE_SIGN_IN_BUTTON'
}
