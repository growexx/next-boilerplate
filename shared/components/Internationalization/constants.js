import { ROUTES } from '@shared/constants'

export const LanguageItems = [
  {
    to: ROUTES.EN,
    tabName: 'EN'
  },
  {
    to: ROUTES.FR,
    tabName: 'FR'
  }
]
