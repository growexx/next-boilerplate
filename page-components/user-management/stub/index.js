import { DEFAULT_LIMIT, DEFAULT_PAGE } from '@shared/constants'

let USERS = [
  {
    firstName: 'Serena',
    lastName: 'Juckes',
    email: 'sjuckes0@forbes.com'
  },
  {
    firstName: 'Trevor',
    lastName: 'Shernock',
    email: 'tshernock1@google.nl'
  }
].map((user, i) => {
  const updatedUser = user
  updatedUser.profileUrl = 'https://thispersondoesnotexist.com/image'
  updatedUser.lastAccessDate = new Date()
  updatedUser.verified = true
  updatedUser.status = i % 2 === 0 ? 'Active' : 'Suspended'
  updatedUser.id = i + 1

  return updatedUser
})

const getSortedUsers = (sortType, sortKey, allUsers) => {
  const updatedSortKey = sortKey
  try {
    return allUsers.sort((user1, user2) => {
      if (updatedSortKey === 'id') {
        return sortType === 1 ? user1[updatedSortKey] - user2[updatedSortKey] : user2[updatedSortKey] - user1[updatedSortKey]
      }

      return sortType === 1
        ? user1[updatedSortKey].localeCompare(user2[updatedSortKey])
        : user2[updatedSortKey].localeCompare(user1[updatedSortKey])
    })
  } catch (error) {
    return allUsers
  }
}

export const getUsersAPIMock = ({ limit = DEFAULT_LIMIT, skip = DEFAULT_PAGE, sortType, sortKey, search, status } = {}) => {
  const users = []
  const skipIndexs = (skip - 1) * limit
  let allUsers = USERS
  allUsers = getSortedUsers(sortType, sortKey, allUsers)

  const contains = (parentObject, key, term) => parentObject[key].toLocaleLowerCase().includes(term.toLocaleLowerCase())

  if (search) {
    allUsers = allUsers.filter(
      (user) => contains(user, 'firstName', search) || contains(user, 'lastName', search) || contains(user, 'email', search)
    )
  }

  if (status) {
    allUsers = allUsers.filter((user) => user.status === status)
  }
  // eslint-disable-next-line no-plusplus
  for (let index = skipIndexs; index < allUsers.length; index++) {
    if (users.length === limit) {
      break
    }

    users.push(allUsers[index])
  }

  return Promise.resolve({
    data: users,
    pagination: {
      pageSize: limit,
      current: skip,
      total: allUsers.length
    },
    status: 1
  })
}

export const deleteUserAPIMock = (id) => {
  USERS = USERS.filter((user) => user.id !== id)

  return Promise.resolve({
    status: 1,
    message: 'Deleted'
  })
}

export const updateUserAPIMock = (payload, newRecord) => {
  if (newRecord) {
    USERS.push({
      ...payload,
      id: USERS.length + 1
    })
  } else {
    USERS = USERS.map((user) => {
      let updatedUser = user
      if (updatedUser.id === payload.id) {
        updatedUser = { ...user, ...payload }
      }

      return updatedUser
    })
  }
  return Promise.resolve({
    status: 1,
    message: 'Success'
  })
}

export const USER_DATA = {
  EMAIL: 'it@growexx.com',
  URL: 'https://growexx.atlassian.net',
  NAME: 'GrowExx',
  _ID: '60631fbe96e98c33713f7000'
}

export const responseWithZeroList = () => ({
  data: [],
  pagination: {
    pageSize: DEFAULT_LIMIT,
    current: DEFAULT_PAGE,
    total: 0
  },
  status: 1
})

export const responseWithList = ({ status } = {}) => ({
  data: USERS.splice(0, DEFAULT_LIMIT),
  pagination: {
    pageSize: DEFAULT_LIMIT,
    current: DEFAULT_PAGE,
    total: USERS.length
  },
  status: status || 1
})

export const failedResponse = {
  data: null,
  status: 0,
  message: 'You do not have access'
}

// Add | Edit
export const addNewUserFailure = () => ({
  response: {}
})

export const addNewUserSuccess = () => ({
  message: 'Done',
  status: 1
})
