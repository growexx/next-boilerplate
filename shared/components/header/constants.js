import React from 'react'
import { UserOutlined } from '@ant-design/icons'

import { ROUTES } from '@shared/constants'

export const MenuItems = [
  {
    to: ROUTES.PROFILE,
    tabName: 'Profile',
    icon: <UserOutlined />
  },
  {
    to: ROUTES.CHANGE_PASSWORD,
    tabName: 'Change Password',
    icon: <UserOutlined />
  }
]
