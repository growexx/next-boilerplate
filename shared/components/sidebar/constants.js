import React from 'react'
import {
  UserOutlined,
  SmileOutlined,
  LoadingOutlined,
  ExportOutlined,
  NumberOutlined,
  FormOutlined,
  LockOutlined,
  PieChartOutlined,
  MessageOutlined,
  ShoppingOutlined,
  SearchOutlined,
  CloudServerOutlined,
  GlobalOutlined
} from '@ant-design/icons'

import { ROUTES } from '@shared/constants'
// import { ROUTES, ROLE_BASED_SIDEBAR_MENU, ROLES, RESTRICTED_ROUTES } from 'containers/constants'

export const MenuItems = [
  {
    to: ROUTES.HOME,
    tabName: 'Home',
    icon: <SmileOutlined />
  },
  {
    to: ROUTES.SAMPLE_FORM,
    tabName: 'Redux-Saga Form',
    icon: <FormOutlined />
  },
  {
    to: ROUTES.LOADER,
    tabName: 'Loader Demo',
    icon: <LoadingOutlined />
  },
  {
    to: ROUTES.EXPORT_DATA,
    tabName: 'Export Data To CSV',
    icon: <ExportOutlined />
  },
  {
    to: ROUTES.NUMERAL_CONVERTER,
    tabName: 'Number Conversion Demo',
    icon: <NumberOutlined />
  },
  {
    to: ROUTES.USERS,
    tabName: 'User Management',
    icon: <UserOutlined />
  },
  {
    to: ROUTES.GITHUB_SEARCH,
    tabName: 'Github Search',
    icon: <SearchOutlined />
  },
  {
    to: ROUTES.TEST_ADMIN_PAGE,
    tabName: 'Admin Page',
    icon: <LockOutlined />
  },
  {
    to: ROUTES.CHARTS,
    tabName: 'Charts',
    icon: <PieChartOutlined />
  },
  {
    to: ROUTES.MULTI_TAB_SUPPORT,
    tabName: 'Multi Tab Communication',
    icon: <MessageOutlined />
  },
  {
    to: ROUTES.PRODUCTS,
    tabName: 'Products',
    icon: <ShoppingOutlined />
  },
  {
    to: ROUTES.SSR_DEMO,
    tabName: 'SSR Demo',
    icon: <CloudServerOutlined />
  },
  {
    to: ROUTES.INTERNATIONALIZATION_DEMO,
    tabName: 'Internationalization Demo',
    icon: <GlobalOutlined />
  }
]

/**
 * Filters Sidebar menu based on role
 * @param {string} role
 */
// export const GET_FILTERED_MENU_ITEM = role =>
//   MenuItems.filter(item =>
//     // Check if route is private then role has access to it
//     RESTRICTED_ROUTES.includes(item.to)
//       ? ROLE_BASED_SIDEBAR_MENU[role || ROLES.USER].includes(item.to)
//       : true,
//   );
// TODO: add user based menu
export const GET_FILTERED_MENU_ITEM = (role) => MenuItems
