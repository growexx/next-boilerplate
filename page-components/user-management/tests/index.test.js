import { fireEvent, render, screen, waitFor } from '@testing-library/react'

import { addNewUserFailure, addNewUserSuccess, failedResponse, responseWithList, responseWithZeroList, USER_DATA } from '../stub'
import { TEST_IDS } from '../constants'
import UserManagement from '../index'
import request from '@shared/utils/request'

jest.mock('@shared/utils/request')

const fieldUpdateViaPlaceHolder = [
  {
    key: 'john.doe@growexx.com',
    value: USER_DATA.EMAIL
  },
  {
    key: 'John',
    value: USER_DATA.NAME
  },
  {
    key: 'Doe',
    value: USER_DATA.NAME
  }
]

describe('Check component:<Users /> is rendering properly', () => {
  beforeEach(() => {
    request.mockImplementation(() => Promise.resolve(responseWithZeroList()))
  })

  afterEach(() => {
    request.mockClear()
  })

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild }
    } = render(<UserManagement />)
    expect(firstChild).toMatchSnapshot()
  })

  it.only('Click: New User Modal should show modal', async () => {
    render(<UserManagement />)
    await waitFor(() => screen.findByTestId(TEST_IDS.ADD_USER))
    fireEvent.click(screen.getByTestId(TEST_IDS.ADD_USER))
    expect(screen.getByText('Add User')).toBeTruthy()
    fireEvent.click(screen.getByTestId(TEST_IDS.ADD_USER))
    fireEvent.click(screen.getByTestId(TEST_IDS.USER_MODAL_CANCEL))
  })

  it('Click Delete: Show Confirmation Modal', async () => {
    render(<UserManagement />)
    await waitFor(() => screen.findByText('Active'))
    fireEvent.click(screen.getByTestId(TEST_IDS.DELETE_BUTTON))
    expect(screen.getByText('OK', { trim: true })).toBeTruthy()
  })

  it('Click Delete: Show Confirmation Modal and click confirm', async () => {
    render(<UserManagement />)
    await waitFor(() => screen.findByText('Active'))
    fireEvent.click(screen.getByTestId(TEST_IDS.DELETE_BUTTON))
    expect(screen.getByText('OK', { trim: true })).toBeTruthy()
    fireEvent.click(screen.getByTestId(TEST_IDS.DELETE_BUTTON_CONFIRMED))
  })
})

describe('Check listing of users is rendering properly', () => {
  beforeEach(() => {
    request.mockImplementation(() => Promise.resolve(responseWithZeroList()))
  })

  afterEach(() => {
    request.mockClear()
  })

  it('No Records found for users', async () => {
    request.mockImplementation(() => Promise.resolve(responseWithZeroList()))
    render(<UserManagement />)
    await waitFor(() => screen.findByText('No data'))
    expect(screen.getByText('No data')).toBeTruthy()
  })

  it('Users Listing with few records should be shown', async () => {
    request.mockImplementation(() => Promise.resolve(responseWithList()))
    render(<UserManagement />)
    await waitFor(() => screen.findByText('Active'))
    expect(screen.getByText('Active')).toBeTruthy()
  })

  it('Users Listing with pagination', async () => {
    request.mockImplementation(() => Promise.resolve(responseWithList()))
    render(<UserManagement />)
    await waitFor(() => screen.findByText('Active'))
    // Expand Data
    fireEvent.click(document.querySelectorAll('.ant-table-row-expand-icon')[0])

    expect(screen.getByText('Active')).toBeTruthy()
    fireEvent.click(screen.getByTitle('2'))
  })

  it('Failed Users Listing api', async () => {
    request.mockImplementationOnce(() => Promise.reject(failedResponse))
    render(<UserManagement />)
    await waitFor(() => screen.findByText('No Data'))
    expect(screen.getByText('No Data')).toBeTruthy()
  })

  it('Toggle User Status', async () => {
    request.mockImplementation(() => Promise.resolve(responseWithList()))
    render(<UserManagement />)
    await waitFor(() => screen.findByText('Active'))
    fireEvent.click(screen.getByTestId(TEST_IDS.STATUS_TOGGLE))
  })

  it('Click Delete: Show Confirmation Modal and click confirm', async () => {
    request.mockImplementation(() => Promise.resolve(failedResponse()))
    render(<UserManagement />)
    await waitFor(() => screen.findByText('Active'))
    fireEvent.click(screen.getByTestId(TEST_IDS.DELETE_BUTTON))
    expect(screen.getByText('OK', { trim: true })).toBeTruthy()
    fireEvent.click(screen.getByTestId(TEST_IDS.DELETE_BUTTON_CONFIRMED))
  })

  it('Click Delete: Show Confirmation Modal and click confirm', async () => {
    request.mockImplementation(() => Promise.resolve(failedResponse()))
    render(<UserManagement />)
    await waitFor(() => screen.findByText('Active'))
    fireEvent.click(screen.getByTestId(TEST_IDS.DELETE_BUTTON))
    expect(screen.getByText('OK', { trim: true })).toBeTruthy()
    fireEvent.click(screen.getByTestId(TEST_IDS.DELETE_CONFIRMATION_CANCEL))
  })

  it('Toggle User Status Local', async () => {
    request.mockImplementation(() => Promise.resolve(responseWithList()))
    render(<UserManagement />)
    await waitFor(() => screen.findByText('Active'))
    fireEvent.click(screen.getByTestId(TEST_IDS.STATUS_TOGGLE))
  })
})

describe('New Users', () => {
  it('Add new users with success', () => {
    request.mockImplementation(() => Promise.resolve(addNewUserSuccess()))
    render(<UserManagement />)
    fireEvent.click(screen.getByTestId(TEST_IDS.ADD_USER))
    fieldUpdateViaPlaceHolder.forEach((d) => {
      fireEvent.change(screen.getByPlaceholderText(d.key), {
        target: { value: d.value }
      })
    })
    expect(screen.getByText('Add User')).toBeTruthy()
    fireEvent.click(screen.getByText('Add'))
  })

  it('Add new user with cancel option', () => {
    request.mockImplementation(() => Promise.resolve(addNewUserSuccess()))
    render(<UserManagement />)
    fireEvent.click(screen.getByTestId(TEST_IDS.ADD_USER))
    fireEvent.click(screen.getByTestId(TEST_IDS.USER_MODAL_CANCEL))
  })

  it('Add new user with failure', () => {
    request.mockImplementation(() => Promise.reject(addNewUserFailure()))
    render(<UserManagement />)
    fireEvent.click(screen.getByTestId(TEST_IDS.ADD_USER))
    fieldUpdateViaPlaceHolder.forEach((d) => {
      fireEvent.change(screen.getByPlaceholderText(d.key), {
        target: { value: d.value }
      })
    })
    expect(screen.getByText('Add User')).toBeTruthy()
    fireEvent.click(screen.getByTestId(TEST_IDS.USER_MODAL_OK))
  })
})

describe('Update User', () => {
  it('Update user with success', async () => {
    request.mockImplementationOnce(() => Promise.resolve(responseWithList()))
    render(<UserManagement />)
    expect(request).toHaveBeenCalledTimes(1)
    await waitFor(() => screen.findByText('Active'))
    fireEvent.click(screen.getByTestId(TEST_IDS.EDIT_BUTTON))
    fieldUpdateViaPlaceHolder.forEach((d) => {
      fireEvent.change(screen.getByPlaceholderText(d.key), {
        target: { value: d.value }
      })
    })
    expect(screen.getByText('Update')).toBeTruthy()
    fireEvent.click(screen.getByTestId(TEST_IDS.ADD_USER))
  })

  it('Update user with cancel', async () => {
    request.mockImplementationOnce(() => Promise.resolve(responseWithList()))
    render(<UserManagement />)
    expect(request).toHaveBeenCalledTimes(1)
    await waitFor(() => screen.findByText('Active'))
    fireEvent.click(screen.getByTestId(TEST_IDS.EDIT_BUTTON))
    expect(screen.getByText('Update')).toBeTruthy()
    fireEvent.click(screen.getByTestId(TEST_IDS.USER_MODAL_CANCEL))
  })

  it('User Update failure', async () => {
    request
      .mockImplementationOnce(() => Promise.resolve(responseWithList()))
      .mockImplementationOnce(() => Promise.resolve(addNewUserFailure()))
    render(<UserManagement />)
    expect(request).toHaveBeenCalledTimes(1)
    await waitFor(() => screen.getByText('Active') || screen.getByText('Suspended'))
    fireEvent.click(screen.getByTestId(TEST_IDS.EDIT_BUTTON))
    fieldUpdateViaPlaceHolder.forEach((d) => {
      fireEvent.change(screen.getByPlaceholderText(d.key), {
        target: { value: d.value }
      })
    })
    expect(screen.getByText('Update')).toBeTruthy()
    fireEvent.click(screen.getByText('Update'))
  })
})

describe('Status Filter', () => {
  it('Status Filter', async () => {
    request.mockImplementationOnce(() => Promise.resolve(responseWithList()))
    render(<UserManagement />)
    await waitFor(() => screen.findByRole('combobox'))
    fireEvent.mouseDown(screen.getByRole('combobox'))
    fireEvent.change(screen.getByRole('combobox'), {
      target: {
        value: 'Active'
      }
    })
    fireEvent.click(document.querySelectorAll('.ant-select-item-option-content')[0])
    fireEvent.blur(screen.getByRole('combobox'))
    fireEvent.focus(screen.getByRole('combobox'))
    fireEvent.click(screen.getByText('Name'))
  })
})

describe('Search & Sorting user list', () => {
  it('Sorting user with success', async () => {
    request.mockImplementationOnce(() => Promise.resolve(responseWithList()))
    render(<UserManagement />)
    fireEvent.click(screen.getByText('Name'))
  })
  it('Search user with success', async () => {
    request.mockImplementationOnce(() => Promise.resolve(responseWithList()))
    render(<UserManagement />)
    document.querySelectorAll('.ant-input-affix-wrapper')[0].children[0].value = 'Hello'
    fireEvent.click(document.querySelectorAll('.ant-input-search-button')[0])
  })

  it('Search user with success', async () => {
    render(<UserManagement />)
    fireEvent.change(screen.getByPlaceholderText('Search User'), {
      target: { value: 'a' }
    })
    fireEvent.change(screen.getByPlaceholderText('Search User'), {
      target: { value: '' }
    })
  })

  it('Sorting', async () => {
    render(<UserManagement />)
    fireEvent.click(screen.getByText('Name'))
    fireEvent.click(screen.getByText('Name'))
    fireEvent.click(screen.getByText('Name'))
    fireEvent.click(screen.getByText('User Id'))
    fireEvent.click(screen.getByText('User Id'))
    fireEvent.click(screen.getByText('User Id'))
  })
})
