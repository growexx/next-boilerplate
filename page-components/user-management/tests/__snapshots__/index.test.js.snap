// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check component:<Users /> is rendering properly Should render and match the snapshot 1`] = `
<div>
  <div
    class="sc-gLLvby hTzSHO"
  >
    <div
      class="ant-table-wrapper sc-jTQDnj iRFfOB css-dev-only-do-not-override-1me4733"
    >
      <div
        class="ant-spin-nested-loading css-dev-only-do-not-override-1me4733"
      >
        <div
          class="ant-spin-container"
        >
          <div
            class="ant-table ant-table-empty"
          >
            <div
              class="ant-table-container"
            >
              <div
                class="ant-table-content"
              >
                <table
                  style="table-layout: auto;"
                >
                  <colgroup>
                    <col
                      class="ant-table-expand-icon-col"
                    />
                    <col
                      style="width: 10%;"
                    />
                    <col
                      style="width: 20%;"
                    />
                    <col
                      style="width: 20%;"
                    />
                    <col
                      style="width: 15%;"
                    />
                    <col
                      style="width: 15%;"
                    />
                    <col
                      style="width: 20%;"
                    />
                  </colgroup>
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <td
                        class="ant-table-cell ant-table-row-expand-icon-cell"
                      />
                      <th
                        aria-label="User Id"
                        class="ant-table-cell ant-table-column-has-sorters"
                        scope="col"
                        tabindex="0"
                      >
                        <div
                          class="ant-table-column-sorters"
                        >
                          <span
                            class="ant-table-column-title"
                          >
                            User Id
                          </span>
                          <span
                            class="ant-table-column-sorter ant-table-column-sorter-full"
                          >
                            <span
                              aria-hidden="true"
                              class="ant-table-column-sorter-inner"
                            >
                              <span
                                aria-label="caret-up"
                                class="anticon anticon-caret-up ant-table-column-sorter-up"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="caret-up"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="0 0 1024 1024"
                                  width="1em"
                                >
                                  <path
                                    d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                  />
                                </svg>
                              </span>
                              <span
                                aria-label="caret-down"
                                class="anticon anticon-caret-down ant-table-column-sorter-down"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="caret-down"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="0 0 1024 1024"
                                  width="1em"
                                >
                                  <path
                                    d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </span>
                        </div>
                      </th>
                      <th
                        aria-label="Name"
                        class="ant-table-cell ant-table-column-has-sorters"
                        scope="col"
                        tabindex="0"
                      >
                        <div
                          class="ant-table-column-sorters"
                        >
                          <span
                            class="ant-table-column-title"
                          >
                            Name
                          </span>
                          <span
                            class="ant-table-column-sorter ant-table-column-sorter-full"
                          >
                            <span
                              aria-hidden="true"
                              class="ant-table-column-sorter-inner"
                            >
                              <span
                                aria-label="caret-up"
                                class="anticon anticon-caret-up ant-table-column-sorter-up"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="caret-up"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="0 0 1024 1024"
                                  width="1em"
                                >
                                  <path
                                    d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                  />
                                </svg>
                              </span>
                              <span
                                aria-label="caret-down"
                                class="anticon anticon-caret-down ant-table-column-sorter-down"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="caret-down"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="0 0 1024 1024"
                                  width="1em"
                                >
                                  <path
                                    d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </span>
                        </div>
                      </th>
                      <th
                        aria-label="Email"
                        class="ant-table-cell ant-table-column-has-sorters"
                        scope="col"
                        tabindex="0"
                      >
                        <div
                          class="ant-table-column-sorters"
                        >
                          <span
                            class="ant-table-column-title"
                          >
                            Email
                          </span>
                          <span
                            class="ant-table-column-sorter ant-table-column-sorter-full"
                          >
                            <span
                              aria-hidden="true"
                              class="ant-table-column-sorter-inner"
                            >
                              <span
                                aria-label="caret-up"
                                class="anticon anticon-caret-up ant-table-column-sorter-up"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="caret-up"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="0 0 1024 1024"
                                  width="1em"
                                >
                                  <path
                                    d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                  />
                                </svg>
                              </span>
                              <span
                                aria-label="caret-down"
                                class="anticon anticon-caret-down ant-table-column-sorter-down"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="caret-down"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="0 0 1024 1024"
                                  width="1em"
                                >
                                  <path
                                    d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </span>
                        </div>
                      </th>
                      <th
                        aria-label="Status"
                        class="ant-table-cell ant-table-column-has-sorters"
                        scope="col"
                        tabindex="0"
                      >
                        <div
                          class="ant-table-column-sorters"
                        >
                          <span
                            class="ant-table-column-title"
                          >
                            Status
                          </span>
                          <span
                            class="ant-table-column-sorter ant-table-column-sorter-full"
                          >
                            <span
                              aria-hidden="true"
                              class="ant-table-column-sorter-inner"
                            >
                              <span
                                aria-label="caret-up"
                                class="anticon anticon-caret-up ant-table-column-sorter-up"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="caret-up"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="0 0 1024 1024"
                                  width="1em"
                                >
                                  <path
                                    d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                  />
                                </svg>
                              </span>
                              <span
                                aria-label="caret-down"
                                class="anticon anticon-caret-down ant-table-column-sorter-down"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="caret-down"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="0 0 1024 1024"
                                  width="1em"
                                >
                                  <path
                                    d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </span>
                        </div>
                      </th>
                      <th
                        class="ant-table-cell"
                        scope="col"
                      >
                        Last Access Date
                      </th>
                      <th
                        class="ant-table-cell"
                        scope="col"
                      >
                        Actions
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    class="ant-table-tbody"
                  >
                    <tr
                      class="ant-table-placeholder"
                    >
                      <td
                        class="ant-table-cell"
                        colspan="7"
                      >
                        <div
                          class="css-dev-only-do-not-override-1me4733 ant-empty ant-empty-normal"
                        >
                          <div
                            class="ant-empty-image"
                          >
                            <svg
                              height="41"
                              viewBox="0 0 64 41"
                              width="64"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <g
                                fill="none"
                                fill-rule="evenodd"
                                transform="translate(0 1)"
                              >
                                <ellipse
                                  cx="32"
                                  cy="33"
                                  fill="#f5f5f5"
                                  rx="32"
                                  ry="7"
                                />
                                <g
                                  fill-rule="nonzero"
                                  stroke="#d9d9d9"
                                >
                                  <path
                                    d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                  />
                                  <path
                                    d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                    fill="#fafafa"
                                  />
                                </g>
                              </g>
                            </svg>
                          </div>
                          <div
                            class="ant-empty-description"
                          >
                            No data
                          </div>
                        </div>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
